### PR TITLE
[Bug][Connector-v2][PulsarSource]Fix pulsar option topic-pattern bug.

### PR DESCRIPTION
**Pull request by lightzhao**
## Purpose of this pull request
When the user does not set the topic-pattern option, it should be judged whether there is a setting, otherwise the following exception will appear.
image

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by TaoZex** [comment]
Is topic-pattern a required parameter?

**Comment by lightzhao** [comment]
> Is topic-pattern a required parameter?

It's not required parameter.

**Review by EricJoy2048 — DISMISSED** [review]

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java [hunk 1]
```diff
@@ -213,12 +213,14 @@ private void setPartitionDiscoverer(Config config) {
...
-            if (this.partitionDiscoverer != null) {
-                throw new PulsarConnectorException(SeaTunnelAPIErrorCode.OPTION_VALIDATION_FAILED, String.format("The properties '%s' and '%s' is exclusive.", TOPIC.key(), TOPIC_PATTERN.key()));
+        if (config.hasPath(TOPIC_PATTERN.key())) {
+            String topicPattern = config.getString(TOPIC_PATTERN.key());
+            if (StringUtils.isNotBlank(topicPattern)) {
+                if (this.partitionDiscoverer != null) {
+                    throw new PulsarConnectorException(SeaTunnelAPIErrorCode.OPTION_VALIDATION_FAILED, String.format("The properties '%s' and '%s' is exclusive.", TOPIC.key(), TOPIC_PATTERN.key()));
+                }
+                this.partitionDiscoverer = new TopicPatternDiscoverer(Pattern.compile(topicPattern));
             }
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: If these two parameters are mutually exclusive, we should prompt the user during the check phase, not here.

**Review by Hisoka-X — DISMISSED** [review]

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java (hunk 1 above)
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: I don't understand what you mean. I just judge whether the topic-pattern parameter is configured.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java [hunk 2]
```diff
@@ -213,12 +213,14 @@ private void setPartitionDiscoverer(Config config) {
...
-        String topicPattern = config.getString(TOPIC_PATTERN.key());
-        if (StringUtils.isNotBlank(topicPattern)) {
-            if (this.partitionDiscoverer != null) {
-                throw new PulsarConnectorException(SeaTunnelAPIErrorCode.OPTION_VALIDATION_FAILED, String.format("The properties '%s' and '%s' is exclusive.", TOPIC.key(), TOPIC_PATTERN.key()));
+        if (config.hasPath(TOPIC_PATTERN.key())) {
+            String topicPattern = config.getString(TOPIC_PATTERN.key());
+            if (StringUtils.isNotBlank(topicPattern)) {
+                if (this.partitionDiscoverer != null) {
+                    throw new PulsarConnectorException(SeaTunnelAPIErrorCode.OPTION_VALIDATION_FAILED, String.format("The properties '%s' and '%s' is exclusive.", TOPIC.key(), TOPIC_PATTERN.key()));
+                }
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: Is this redundant?

Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: it needs.

Comment (CalvinKirs) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: I mean, if these two parameters are mutually exclusive, then they should be verified in the check phase, not here. so it is redundant.

**Review by CalvinKirs — CHANGES_REQUESTED** [review]

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java (hunk 2 above)
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: I think the judgment here also belongs to the check phase.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java [hunk 3]
```diff
@@ -213,12 +213,14 @@ private void setPartitionDiscoverer(Config config) {
         if (StringUtils.isNotBlank(topic)) {
             this.partitionDiscoverer = new TopicListDiscoverer(Arrays.asList(StringUtils.split(topic, ",")));
         }
-        String topicPattern = config.getString(TOPIC_PATTERN.key());
-        if (StringUtils.isNotBlank(topicPattern)) {
-            if (this.partitionDiscoverer != null) {
-                throw new PulsarConnectorException(SeaTunnelAPIErrorCode.OPTION_VALIDATION_FAILED, String.format("The properties '%s' and '%s' is exclusive.", TOPIC.key(), TOPIC_PATTERN.key()));
+        if (config.hasPath(TOPIC_PATTERN.key())) {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java:218]: If `TOPIC_PATTERN` and `TOPIC` are exclusive, you need add `exclusive(TOPIC_PATTERN, TOPIC)` in OptionRule. And you can reference https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkFactory.java

Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java:218]: The exclusive configuration of these two parameters has been added.
image

**Comment by lightzhao** [comment]
@EricJoy2048 @Hisoka-X PTAL

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java (hunk 3 above)
Comment (CalvinKirs) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java:218]: back to this question. https://github.com/apache/incubator-seatunnel/pull/3989#discussion_r1090297180

Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java:218]: ok, redundancy judgment deleted.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java [hunk 4]
```diff
@@ -213,12 +213,11 @@ private void setPartitionDiscoverer(Config config) {
         if (StringUtils.isNotBlank(topic)) {
             this.partitionDiscoverer = new TopicListDiscoverer(Arrays.asList(StringUtils.split(topic, ",")));
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: `topic` should use `config.hasPath(TOPIC_PATTERN.key()` judge whether there is

Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: done.

**Review by CalvinKirs — APPROVED** [review]

**Review by Hisoka-X — APPROVED** [review]

**Pull request merged**